lib/linkerd2-proxy-discover/src/buffer.rs

```diff
@@ -0,0 +1,161 @@
+    future: F,
+    capacity: usize,
+    _marker: std::marker::PhantomData<fn() -> D>,
+}
+
+pub struct Daemon<D: discover::Discover> {
+    discover: D,
+    disconnect_rx: oneshot::Receiver<Never>,
+    tx: mpsc::Sender<discover::Change<D::Key, D::Service>>,
+}
+
+#[derive(Clone, Debug)]
+pub struct Lost(());
+
+impl<M> Buffer<M> {
+    pub fn new<T, D>(capacity: usize, inner: M) -> Self
+    where
+        T: Clone + std::fmt::Display,
+        M: tower::Service<T, Response = D>,
+        D: discover::Discover + Send + 'static,
+        D::Error: Into<Error>,
+        D::Key: Send,
+        D::Service: Send,
+    {
+        Self { capacity, inner }
+    }
+}
+
+impl<T, M, D> tower::Service<T> for Buffer<M>
+where
+    T: Clone + std::fmt::Display,
+    M: tower::Service<T, Response = D>,
+    D: discover::Discover + Send + 'static,
+    D::Error: Into<Error>,
+    D::Key: Send,
+    D::Service: Send,
+{
+    type Response = Discover<D::Key, D::Service>;
+    type Error = M::Error;
+    type Future = DiscoverFuture<T, M::Future, M::Response>;
+
+    fn poll_ready(&mut self) -> Poll<(), Self::Error> {
+        self.inner.poll_ready()
+    }
+
+    fn call(&mut self, target: T) -> Self::Future {
+        let future = self.inner.call(target.clone());
+        Self::Future {
+            future,
+            target,
+            capacity: self.capacity,
+            _marker: std::marker::PhantomData,
+        }
+    }
+}
+
+impl<T, F, D> Future for DiscoverFuture<T, F, D>
+where
+    T: std::fmt::Display,
+    F: Future<Item = D>,
+    D: discover::Discover + Send + 'static,
+    D::Error: Into<Error>,
+    D::Key: Send,
+    D::Service: Send,
+{
+    type Item = Discover<D::Key, D::Service>;
+    type Error = F::Error;
+
+    fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
+        let discover = try_ready!(self.future.poll());
+
+        let (tx, rx) = mpsc::channel(self.capacity);
+        let (_disconnect_tx, disconnect_rx) = oneshot::channel();
+        let daemon = Daemon {
+            discover,
+            disconnect_rx,
+            tx,
+        }
+        .instrument(info_span!("discover", target = %self.target));
+        task::spawn(daemon);
+
+        Ok(Discover { rx, _disconnect_tx }.into())
+    }
+}
+
+impl<D> Future for Daemon<D>
+where
+    D: discover::Discover,
+    D::Error: Into<Error>,
+{
+    type Item = ();
+    type Error = ();
+
+    fn poll(&mut self) -> Poll<(), ()> {
+        loop {
+            match self.disconnect_rx.poll() {
+                Ok(Async::NotReady) => {}
+                Err(_lost) => return Ok(().into()),
+                Ok(Async::Ready(n)) => match n {},
+            }
+
+            try_ready!(self
+                .tx
+                .poll_ready()
+                .map_err(|_| tracing::trace!("lost sender")));
+
+            let up = try_ready!(self.discover.poll().map_err(|e| {
+                let e: Error = e.into();
+                tracing::debug!("resoution lost: {}", e);
```

Comment:
Typo:
```suggestion
                tracing::debug!("resolution lost: {}", e);
```

Reply:
Also, nit: prefer structured fields:
```suggestion
                tracing::debug!(message = "resoution lost", error = %e);
```

Reply:
in this case, i disagree? this is all one concatenated error message.

Reply:
fair enough